### ISSUE TITLE
[s390x] Fails to compile against recent glibc

### ISSUE THREAD

**Issue by directhex** [mono/mono#9009]
```
  CC       libmonoutils_la-os-event-unix.lo
In file included from mono-threads.h:15:0,
                 from mono-lazy-init.h:19,
                 from os-event-unix.c:14:
../../mono/utils/mono-stack-unwinding.h:116:14: error: field 'ctx' has incomplete type
  MonoContext ctx;
              ^~~
```

Presumably this is similar to https://github.com/mono/mono/commit/9cd9e3b0a13e1f0b757606e5966e1ce27722ed95 i.e. `ucontext` was removed from glibc: https://sourceware.org/git/gitweb.cgi?p=glibc.git;a=commitdiff;h=251287734e89a52da3db682a8241eb6bccc050c9;hp=c86ed71d633c22d6f638576f7660c52a5f783d66

Changing `typedef struct ucontext MonoContext` to `typedef ucontext_t MonoContext` fixes the runtime build - but segfaults with no detail when the runtime tries to actually do anything.

Paging @nealef since nobody else will look at it

**Comment by nealef** [comment]
From memory, MoniContext is identical to mcontext or u context and us defined in one of the header files in utils. I'll take a look at it tomorrow or later today if I can get a moment.


-------- Original message --------

**Comment by akoeplinger** [comment]
Interestingly, this seems to have started working again recently: https://jenkins.mono-project.com/job/test-mono-mainline-community/label=centos-s390x/3096/

**Comment by directhex** [comment]
```(sid_s390x-dchroot)directhex@zelenka:~/monoreal$ mono/mini/mono --debug mcs/class/lib/monolite-linux/1051700017/mcs.exe 
Segmentation fault
```

Still sorta boned. Debian Unstable, mono master, glibc 2.27

**Comment by nealef** [comment]
I'll take a look. I wonder if the vector registers are now part of the context and that's causing grief. I don't have access to a Debian system though. Is it building on centos & sles systems?

**Comment by directhex** [comment]
The runtime is compiling right now on the CI machines, the classlib isn't (not sure why).

As far as I can tell, the build failure in Debian started with glibc 2.26, which is newer than what's in CentOS/SLES

**Comment by directhex** [comment]
If it helps:

```
(sid_s390x-dchroot)directhex@zelenka:~/monoreal$ gdb mono/mini/mono
GNU gdb (Debian 7.12-6+b2) 7.12.0.20161007-git
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "s390x-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from mono/mini/mono...done.
(gdb) start mcs/class/lib/monolite-linux/1051700017/mcs.exe
Temporary breakpoint 1 at 0x3e320: file main.c, line 401.
Starting program: /home/directhex/monoreal/mono/mini/mono mcs/class/lib/monolite-linux/1051700017/mcs.exe
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/s390x-linux-gnu/libthread_db.so.1".

Temporary breakpoint 1, main (argc=2, argv=0x3fffffff5a8) at main.c:401
401	{
(gdb) c
Continuing.

Program received signal SIGSEGV, Segmentation fault.
mono_tls_init_runtime_keys () at mono-tls.c:212
212		MONO_THREAD_VAR_OFFSET (mono_tls_thread, tls_offsets [TLS_KEY_THREAD]);
```

**Comment by directhex** [comment]
I'm reasonably sure Debian isn't building with -fPIE, so it'll be using the default definition of MONO_THREAD_VAR_OFFSET

**Comment by directhex** [comment]
I'm wrong.

```
(sid_s390x-dchroot)directhex@zelenka:~/monoreal$ gcc -dM -E - < /dev/null | grep PIE
#define __PIE__ 2
```

**Comment by directhex** [comment]
Also, I'm conflating two different things, and I shouldn't be.

I can't build Mono Master on Debian s390x. That's giving the above segfault.

I also can't build the current Debian version (4.6) on s390x (after changing `ucontext` to `ucontext_t`. That's giving a different segfault:

```
Starting program: /home/directhex/mono-4.6.2.7+dfsg/mono/mini/mono mcs/class/lib/monolite/basic.exe
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/s390x-linux-gnu/libthread_db.so.1".

Temporary breakpoint 1, main (argc=2, argv=0x3fffffff548) at main.c:173
173	{
(gdb) c
Continuing.
[New Thread 0x3fffd3ff910 (LWP 32682)]
[New Thread 0x3fffb77b910 (LWP 32683)]

Thread 1 "Main" received signal SIGSEGV, Segmentation fault.
0x00000001001a06ea in mono_verifier_verify_method_header (image=image@entry=0x10041a790, offset=offset@entry=333887, error_list=error_list@entry=0x0) at metadata-verify.c:4034
4034	{
(gdb) bt
#0  0x00000001001a06ea in mono_verifier_verify_method_header (image=image@entry=0x10041a790, offset=offset@entry=333887, error_list=error_list@entry=0x0) at metadata-verify.c:4034
#1  0x0000000100171fbe in mono_method_get_header_checked (method=0x1004d8618, method@entry=0x0, error=0x101ad6da8, error@entry=<error reading variable: value has been optimized out>) at loader.c:2558
#2  0x00000001000560fc in mono_method_to_ir (cfg=cfg@entry=0x101ad6920, method=method@entry=0x1004d8618, start_bblock=start_bblock@entry=0x0, end_bblock=end_bblock@entry=0x0, return_var=return_var@entry=0x0, inline_args=<optimized out>, 
    inline_offset=<optimized out>, is_virtual_call=<optimized out>) at method-to-ir.c:8203
#3  0x000000010011947c in mini_method_compile (method=method@entry=0x1004d8618, opts=opts@entry=370223487, domain=domain@entry=0x100415f30, flags=flags@entry=JIT_FLAG_RUN_CCTORS, parts=parts@entry=0, aot_method_index=<optimized out>)
    at mini.c:3519
#4  0x000000010011b2c6 in mono_jit_compile_method_inner (method=method@entry=0x1004d8618, target_domain=target_domain@entry=0x100415f30, opt=opt@entry=370223487, error=error@entry=0x3ffff800a98) at mini.c:4200
#5  0x000000010003fd96 in mono_jit_compile_method_with_opt (method=method@entry=0x1004d8618, opt=<optimized out>, error=0x3ffff800a98, error@entry=<error reading variable: value has been optimized out>) at mini-runtime.c:1910
#6  0x00000001000406b2 in mono_jit_compile_method (method=method@entry=0x1004d8618, error=<error reading variable: value has been optimized out>) at mini-runtime.c:1954
#7  0x00000001000d55ac in common_call_trampoline (regs=<optimized out>, regs@entry=0x3ffff800c08, code=code@entry=0x3fffdfd7046 "\343\060\360\240", m=m@entry=0x1004d8618, vt=vt@entry=0x0, vtable_slot=<optimized out>, vtable_slot@entry=0x0, 
    error=0x3ffff800a98) at mini-trampolines.c:702
#8  0x00000001000d5c7c in mono_magic_trampoline (regs=0x3ffff800c08, code=0x3fffdfd7046 "\343\060\360\240", arg=0x1004d8618, tramp=<optimized out>) at mini-trampolines.c:828
#9  0x000003fffdfe60e2 in ?? ()
#10 0x000003fffdfd7046 in ?? ()
PC not saved
```

Changing `ucontext` to `ucontext_t` should be correct, as per the glibc source:

```
--- a/sysdeps/unix/sysv/linux/s390/sys/ucontext.h
+++ b/sysdeps/unix/sysv/linux/s390/sys/ucontext.h
@@ -83,10 +83,10 @@ typedef struct
 #undef __ctx
 
 /* Userlevel context.  */
-typedef struct ucontext
+typedef struct ucontext_t
   {
     unsigned long int uc_flags;
-    struct ucontext *uc_link;
+    struct ucontext_t *uc_link;
     stack_t uc_stack;
     mcontext_t uc_mcontext;
     sigset_t uc_sigmask;
```

But the other segfault issue remains (and I'm aware I'm talking about an ancient version here - newer is blocked on https://github.com/mono/mono/issues/7445 but right now this blocks Mono on all architectures in Debian)

**Comment by nealef** [comment]
Is there a Debian system I could get access to for this?

**Comment by directhex** [comment]
https://dsa.debian.org/doc/guest-account/

I can sponsor the access request

**Comment by aviau** [comment]
This has been blocking mono migrations in Debian for months now. Are there anything we can do to move forward? I am also happy to sponsor a request to a Debian porter machine. Do you want me to move on with a request?

**Comment by marek-safar** [comment]
@nealef did you manage to check this out?

**Comment by nealef** [comment]
@marek-safar It fell off my radar (I was in Australia when the early posts were made). @directhex Can I still get that debian account?

**Comment by directhex** [comment]
Yes, it still involves the relevant paperwork, but I don't see why not.

I'm also in the process of trying to make a recent Mono available in Debian's repositories, and s390x still failing to build would be a blocker

**Comment by directhex** [comment]
5.12, with `typedef ucontext_t MonoContext`, works.

**Comment by directhex** [comment]
Er, wow, it was working until it got to mdoc (!) then the runtime exploded

```
MDOC    [net_4_x-linux] cs-errors.tree
Stacktrace:

  at <unknown> <0xffffffff>
  at System.Configuration.ConfigurationElement.get_Properties () [0x00012] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.ElementInformation..ctor (System.Configuration.ConfigurationElement,System.Configuration.PropertyInformation) [0x0001f] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.ConfigurationElement.InitFromProperty (System.Configuration.PropertyInformation) [0x00000] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.ConfigurationElementCollection.InitFromProperty (System.Configuration.PropertyInformation) [0x00057] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.PropertyInformation.get_Value () [0x0002d] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.ConfigurationElement.get_Item (string) [0x0002e] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.ConfigurationElement.get_Item (System.Configuration.ConfigurationProperty) [0x00007] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Diagnostics.TraceSection.get_Listeners () [0x00000] in <78c9c3dcd7b64ce2bd1493515115517a>:0
  at System.Diagnostics.SystemDiagnosticsSection.InitializeDefault () [0x00006] in <78c9c3dcd7b64ce2bd1493515115517a>:0
  at System.Configuration.ConfigurationElement.Reset (System.Configuration.ConfigurationElement) [0x00023] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.Configuration.GetSectionInstance (System.Configuration.SectionInfo,bool) [0x000b8] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.Configuration.GetSectionInstance (System.Configuration.SectionInfo,bool) [0x0005f] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.ConfigurationSectionCollection.get_Item (string) [0x00032] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.Configuration.GetSection (string) [0x00021] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.ClientConfigurationSystem.System.Configuration.Internal.IInternalConfigSystem.GetSection (string) [0x00006] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.ConfigurationManager.GetSection (string) [0x00005] in <c4f58fa43ab34c1db3d70810236c1708>:0
  at System.Configuration.PrivilegedConfigurationManager.GetSection (string) [0x00000] in <78c9c3dcd7b64ce2bd1493515115517a>:0
  at System.Diagnostics.DiagnosticsConfiguration.GetConfigSection () [0x00000] in <78c9c3dcd7b64ce2bd1493515115517a>:0
  at System.Diagnostics.DiagnosticsConfiguration.Initialize () [0x00033] in <78c9c3dcd7b64ce2bd1493515115517a>:0
  at System.Diagnostics.DiagnosticsConfiguration.get_SwitchSettings () [0x00000] in <78c9c3dcd7b64ce2bd1493515115517a>:0
  at System.Diagnostics.Switch.InitializeConfigSettings () [0x00019] in <78c9c3dcd7b64ce2bd1493515115517a>:0
  at System.Diagnostics.Switch.InitializeWithStatus () [0x0004f] in <78c9c3dcd7b64ce2bd1493515115517a>:0
  at System.Diagnostics.Switch.get_SwitchSetting () [0x0000d] in <78c9c3dcd7b64ce2bd1493515115517a>:0
  at System.Diagnostics.BooleanSwitch.get_Enabled () [0x00000] in <78c9c3dcd7b64ce2bd1493515115517a>:0
  at System.Xml.Serialization.TempAssembly.LoadGeneratedAssembly (System.Type,string,System.Xml.Serialization.XmlSerializerImplementation&) [0x0001d] in <c60d2023608448a9862558e4daff55a8>:0
  at System.Xml.Serialization.XmlSerializer..ctor (System.Type,string) [0x00096] in <c60d2023608448a9862558e4daff55a8>:0
  at System.Xml.Serialization.XmlSerializer..ctor (System.Type) [0x00000] in <c60d2023608448a9862558e4daff55a8>:0
  at Monodoc.Providers.ErrorProvider.ReadConfig (string) [0x00000] in <e561899be30442afa6cd92c6752cb5b3>:0
  at Monodoc.Providers.ErrorProvider..ctor (string) [0x00006] in <e561899be30442afa6cd92c6752cb5b3>:0
  at Mono.Documentation.MDocAssembler.<Run>m__1 (string) [0x00000] in <3bdd5bfc0d704af09660d1a9c8996f52>:0
  at System.Linq.Enumerable/SelectListIterator`2<TSource_REF, TResult_REF>.MoveNext () [0x00059] in <ab6537f0f5854bcbb2aaa57fd06b093d>:0
  at System.Collections.Generic.List`1<T_REF>.InsertRange (int,System.Collections.Generic.IEnumerable`1<T_REF>) [0x000ff] in <609c3b2035cf4e378b67cebd3c66589f>:0
  at System.Collections.Generic.List`1<T_REF>.AddRange (System.Collections.Generic.IEnumerable`1<T_REF>) [0x00000] in <609c3b2035cf4e378b67cebd3c66589f>:0
  at Mono.Documentation.MDocAssembler.Run (System.Collections.Generic.IEnumerable`1<string>) [0x002c0] in <3bdd5bfc0d704af09660d1a9c8996f52>:0
  at Mono.Documentation.MDoc.Run (string[]) [0x001f6] in <3bdd5bfc0d704af09660d1a9c8996f52>:0
  at Mono.Documentation.MDoc.Main (string[]) [0x00006] in <3bdd5bfc0d704af09660d1a9c8996f52>:0
  at (wrapper runtime-invoke) <Module>.runtime_invoke_void_object (object,intptr,intptr,intptr) [0x0004e] in <3bdd5bfc0d704af09660d1a9c8996f52>:0
/proc/self/maps:
10b700000-10bb43000 r-xp 00000000 fe:00 1236986                          /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono-sgen
10bb43000-10bb4a000 r--p 00442000 fe:00 1236986                          /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono-sgen
10bb4a000-10bb4f000 rw-p 00449000 fe:00 1236986                          /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono-sgen
10bb4f000-10bb65000 rw-p 00000000 00:00 0
13f154000-13f684000 rw-p 00000000 00:00 0                                [heap]
3ffafa00000-3ffafb00000 rw-p 00000000 00:00 0
3ffafb80000-3ffafbd1000 r--p 00000000 fe:00 686698                       /home/directhex/mono-5.12.0.309+dfsg/mcs/class/lib/net_4_x-linux/Mono.Security.dll
3ffafc00000-3ffafc20000 r--p 00000000 fe:00 686708                       /home/directhex/mono-5.12.0.309+dfsg/mcs/class/lib/net_4_x-linux/System.Configuration.dll
3ffafc80000-3ffaffb0000 r--p 00000000 fe:00 686710                       /home/directhex/mono-5.12.0.309+dfsg/mcs/class/lib/net_4_x-linux/System.Xml.dll
3ffb0000000-3ffb0021000 rw-p 00000000 00:00 0
3ffb0021000-3ffb4000000 ---p 00000000 00:00 0
3ffb4080000-3ffb409f000 r--p 00000000 fe:00 686808                       /home/directhex/mono-5.12.0.309+dfsg/mcs/class/lib/net_4_x-linux/System.Xml.Linq.dll
3ffb4100000-3ffb420b000 r--p 00000000 fe:00 686718                       /home/directhex/mono-5.12.0.309+dfsg/mcs/class/lib/net_4_x-linux/System.Core.dll
3ffb4280000-3ffb42d9000 r--p 00000000 fe:00 686910                       /home/directhex/mono-5.12.0.309+dfsg/mcs/class/lib/net_4_x-linux/Mono.Cecil.dll
3ffb4300000-3ffb45bd000 r--p 00000000 fe:00 686712                       /home/directhex/mono-5.12.0.309+dfsg/mcs/class/lib/net_4_x-linux/System.dll
3ffb45ff000-3ffb4600000 ---p 00000000 00:00 0
3ffb4600000-3ffb4800000 rw-p 00000000 00:00 0
3ffb4800000-3ffb4bee000 r--p 00000000 fe:00 686692                       /home/directhex/mono-5.12.0.309+dfsg/mcs/class/lib/net_4_x-linux/mscorlib.dll
3ffb4bff000-3ffb5bff000 rw-p 00000000 00:00 0
3ffb5bff000-3ffb5c00000 ---p 00000000 00:00 0
3ffb5c00000-3ffb6800000 rw-p 00000000 00:00 0
3ffb6880000-3ffb6990000 r--p 00000000 fe:00 686986                       /home/directhex/mono-5.12.0.309+dfsg/mcs/class/lib/net_4_x-linux/monodoc.dll
3ffb69fc000-3ffb6a7c000 rw-p 00000000 00:00 0
3ffb6a80000-3ffb6aed000 r--p 00000000 fe:00 687249                       /home/directhex/mono-5.12.0.309+dfsg/mcs/class/lib/net_4_x-linux/mdoc.exe
3ffb6b00000-3ffb6c9e000 r-xp 00000000 fe:00 692902                       /lib/s390x-linux-gnu/libc-2.27.so
3ffb6c9e000-3ffb6ca2000 r--p 0019d000 fe:00 692902                       /lib/s390x-linux-gnu/libc-2.27.so
3ffb6ca2000-3ffb6ca4000 rw-p 001a1000 fe:00 692902                       /lib/s390x-linux-gnu/libc-2.27.so
3ffb6ca4000-3ffb6ca8000 rw-p 00000000 00:00 0
3ffb6d00000-3ffb6d1b000 r-xp 00000000 fe:00 693408                       /lib/s390x-linux-gnu/libpthread-2.27.so
3ffb6d1b000-3ffb6d1c000 ---p 0001b000 fe:00 693408                       /lib/s390x-linux-gnu/libpthread-2.27.so
3ffb6d1c000-3ffb6d1d000 r--p 0001b000 fe:00 693408                       /lib/s390x-linux-gnu/libpthread-2.27.so
3ffb6d1d000-3ffb6d1e000 rw-p 0001c000 fe:00 693408                       /lib/s390x-linux-gnu/libpthread-2.27.so
3ffb6d1e000-3ffb6d22000 rw-p 00000000 00:00 0
3ffb6d80000-3ffb6d83000 r-xp 00000000 fe:00 693025                       /lib/s390x-linux-gnu/libdl-2.27.so
3ffb6d83000-3ffb6d84000 r--p 00002000 fe:00 693025                       /lib/s390x-linux-gnu/libdl-2.27.so
3ffb6d84000-3ffb6d85000 rw-p 00003000 fe:00 693025                       /lib/s390x-linux-gnu/libdl-2.27.so
3ffb6d86000-3ffb6de1000 rwxp 00000000 00:00 0
3ffb6e00000-3ffb6e08000 r-xp 00000000 fe:00 693410                       /lib/s390x-linux-gnu/librt-2.27.so
3ffb6e08000-3ffb6e09000 r--p 00007000 fe:00 693410                       /lib/s390x-linux-gnu/librt-2.27.so
3ffb6e09000-3ffb6e0a000 rw-p 00008000 fe:00 693410                       /lib/s390x-linux-gnu/librt-2.27.so
3ffb6e20000-3ffb6e80000 rwxp 00000000 00:00 0
3ffb6e80000-3ffb6f32000 r-xp 00000000 fe:00 693033                       /lib/s390x-linux-gnu/libm-2.27.so
3ffb6f32000-3ffb6f33000 r--p 000b1000 fe:00 693033                       /lib/s390x-linux-gnu/libm-2.27.so
3ffb6f33000-3ffb6f34000 rw-p 000b2000 fe:00 693033                       /lib/s390x-linux-gnu/libm-2.27.so
3ffb6f40000-3ffb6f80000 rwxp 00000000 00:00 0
3ffb6f80000-3ffb6f81000 rw-s 00000000 00:2d 20625160                     /dev/shm/mono.45160
3ffb6f87000-3ffb6fa0000 rwxp 00000000 00:00 0
3ffb6fa0000-3ffb6fa1000 rw-p 00000000 00:00 0
3ffb6fa1000-3ffb7000000 ---p 00000000 00:00 0
3ffb7000000-3ffb7024000 r-xp 00000000 fe:00 692815                       /lib/s390x-linux-gnu/ld-2.27.so
3ffb7024000-3ffb7025000 rw-p 00000000 00:00 0
3ffb7025000-3ffb7026000 r--p 00024000 fe:00 692815                       /lib/s390x-linux-gnu/ld-2.27.so
3ffb7026000-3ffb7028000 rw-p 00025000 fe:00 692815                       /lib/s390x-linux-gnu/ld-2.27.so
3ffb7028000-3ffb7058000 rwxp 00000000 00:00 0
3ffb7058000-3ffb7066000 rw-p 00000000 00:00 0
3ffb7066000-3ffb7076000 rwxp 00000000 00:00 0
3ffb7076000-3ffb7077000 ---p 00000000 00:00 0
3ffb7077000-3ffb7078000 r--p 00000000 00:00 0
3ffb7078000-3ffb707e000 rw-p 00000000 00:00 0
3ffb707e000-3ffb7080000 r-xp 00000000 00:00 0                            [vdso]
3fff3c5e000-3fff3c80000 rw-p 00000000 00:00 0                            [stack]

Native stacktrace:

        /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono(+0xcca82) [0x10b7cca82]
        /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono(+0x3433e) [0x10b73433e]
        [0x3fff3c7a200]
        /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono(mono_metadata_get_inflated_signature+0x32) [0x10b8eb502]
        /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono(+0x1cdd42) [0x10b8cdd42]
        /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono(+0x811b0) [0x10b7811b0]
        /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono(+0x12ced0) [0x10b82ced0]
        /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono(+0x12f096) [0x10b82f096]
        /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono(+0x32aa4) [0x10b732aa4]
        /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono(+0xcff0e) [0x10b7cff0e]
        /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono(+0xd0814) [0x10b7d0814]
        [0x3ffb70660e2]

Debug info from gdb:

[New LWP 45182]
[New LWP 45183]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/s390x-linux-gnu/libthread_db.so.1".
0x000003ffb6d12bfa in waitpid () from /lib/s390x-linux-gnu/libpthread.so.0
  Id   Target Id         Frame
* 1    Thread 0x3ffb7078730 (LWP 45160) "mono" 0x000003ffb6d12bfa in waitpid () from /lib/s390x-linux-gnu/libpthread.so.0
  2    Thread 0x3ffb63ff910 (LWP 45182) "SGen worker" 0x000003ffb6d0db42 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/s390x-linux-gnu/libpthread.so.0
  3    Thread 0x3ffb47ff910 (LWP 45183) "Finalizer" 0x000003ffb6d1077a in do_futex_wait.constprop () from /lib/s390x-linux-gnu/libpthread.so.0

Thread 3 (Thread 0x3ffb47ff910 (LWP 45183)):
#0  0x000003ffb6d1077a in do_futex_wait.constprop () from /lib/s390x-linux-gnu/libpthread.so.0
#1  0x000003ffb6d1088c in __new_sem_wait_slow.constprop.0 () from /lib/s390x-linux-gnu/libpthread.so.0
#2  0x000000010b98089e in mono_os_sem_wait (flags=MONO_SEM_FLAGS_ALERTABLE, sem=0x10bb53af0 <finalizer_sem>) at ../../mono/utils/mono-os-semaphore.h:209
#3  mono_coop_sem_wait (flags=MONO_SEM_FLAGS_ALERTABLE, sem=0x10bb53af0 <finalizer_sem>) at ../../mono/utils/mono-coop-semaphore.h:43
#4  finalizer_thread (unused=unused@entry=0x0) at gc.c:893
#5  0x000000010b92da12 in start_wrapper_internal (stack_ptr=0x3ffb4800000, start_info=0x0) at threads.c:1073
#6  start_wrapper (data=0x13f1d4450) at threads.c:1133
#7  0x000003ffb6d07b4e in start_thread () from /lib/s390x-linux-gnu/libpthread.so.0
#8  0x000003ffb6bfe366 in ?? () from /lib/s390x-linux-gnu/libc.so.6

Thread 2 (Thread 0x3ffb63ff910 (LWP 45182)):
#0  0x000003ffb6d0db42 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/s390x-linux-gnu/libpthread.so.0
#1  0x000000010b9ed296 in mono_os_cond_wait (mutex=0x10bb62758 <lock>, cond=0x10bb62728 <work_cond>) at ../../mono/utils/mono-os-mutex.h:173
#2  get_work (job=<synthetic pointer>, do_idle=<synthetic pointer>, work_context=<synthetic pointer>, worker_index=<optimized out>) at sgen-thread-pool.c:165
#3  thread_func (data=<optimized out>) at sgen-thread-pool.c:196
#4  0x000003ffb6d07b4e in start_thread () from /lib/s390x-linux-gnu/libpthread.so.0
#5  0x000003ffb6bfe366 in ?? () from /lib/s390x-linux-gnu/libc.so.6

Thread 1 (Thread 0x3ffb7078730 (LWP 45160)):
#0  0x000003ffb6d12bfa in waitpid () from /lib/s390x-linux-gnu/libpthread.so.0
#1  0x000000010b7ccb40 in mono_handle_native_crash (signal=signal@entry=0x10ba1c12e "SIGSEGV", ctx=ctx@entry=0x3fff3c7a288, info=info@entry=0x3fff3c7a208) at mini-exceptions.c:3032
#2  0x000000010b73433e in mono_sigsegv_signal_handler (_dummy=<optimized out>, _info=0x3fff3c7a208, context=0x3fff3c7a288) at mini-runtime.c:3164
#3  <signal handler called>
#4  mono_metadata_get_inflated_signature (sig=sig@entry=0x13f5ac450, context=context@entry=0x3f3851ec00000000) at metadata.c:3079
#5  0x000000010b8cdd42 in mono_method_get_signature_checked (method=method@entry=0x13f5a1d80, image=0x13f3f32a0, image@entry=0x13f5a94c0, token=token@entry=167772188, context=0x3f3851ec00000000, context@entry=<error reading variable: value has been optimized out>, error=error@entry=0x13f5ac040) at loader.c:733
#6  0x000000010b7811b0 in mono_method_to_ir (cfg=cfg@entry=0x13f5abbb0, method=method@entry=0x13f5a1d40, start_bblock=<optimized out>, start_bblock@entry=0x0, end_bblock=<optimized out>, end_bblock@entry=0x0, return_var=return_var@entry=0x0, inline_args=<optimized out>, inline_offset=<optimized out>, is_virtual_call=<optimized out>) at method-to-ir.c:8432
#7  0x000000010b82ced0 in mini_method_compile (method=method@entry=0x13f5a1d40, opts=opts@entry=370223487, domain=domain@entry=0x13f162b40, flags=flags@entry=JIT_FLAG_RUN_CCTORS, parts=parts@entry=0, aot_method_index=<optimized out>) at mini.c:3455
#8  0x000000010b82f096 in mono_jit_compile_method_inner (method=method@entry=0x13f5a1d40, target_domain=target_domain@entry=0x13f162b40, opt=opt@entry=370223487, error=0x3fff3c7b390, error@entry=<error reading variable: value has been optimized out>) at mini.c:4190
#9  0x000000010b732aa4 in mono_jit_compile_method_with_opt (method=method@entry=0x13f5a1d40, opt=<optimized out>, jit_only=jit_only@entry=0, error=<optimized out>, error@entry=0x3fff3c7b390) at mini-runtime.c:2150
#10 0x000000010b733a36 in mono_jit_compile_method (method=method@entry=0x13f5a1d40, error=error@entry=0x3fff3c7b390) at mini-runtime.c:2196
#11 0x000000010b7cff0e in common_call_trampoline (regs=regs@entry=0x3fff3c7b500, code=code@entry=0x3ffb6dde556 "\343\060\360\240", m=m@entry=0x13f5a1d40, vt=vt@entry=0x0, vtable_slot=<optimized out>, vtable_slot@entry=0x0, error=0x3fff3c7b390) at mini-trampolines.c:736
#12 0x000000010b7d0814 in mono_magic_trampoline (regs=0x3fff3c7b500, code=0x3ffb6dde556 "\343\060\360\240", arg=0x13f5a1d40, tramp=<optimized out>) at mini-trampolines.c:867
#13 0x000003ffb70660e2 in ?? ()
#14 0x000003ffb6dde556 in ?? ()
#15 0x000003ffb6ddb24e in ?? ()
/tmp/mono-gdb-commands.t0K0S6:3: Error in sourced command file:
PC not saved

=================================================================
Got a SIGSEGV while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================

Aborted
make[8]: *** [Makefile:149: cs-errors.tree] Error 134
```

:man_shrugging:

**Comment by nealef** [comment]
That context value looks suspect.




#4  mono_metadata_get_inflated_signature (sig=sig@entry=0x13f5ac450, context=context@entry=0x3f3851ec00000000) at metadata.c:3079

**Comment by directhex** [comment]
@nealef What would help as a next debugging step?

**Comment by nealef** [comment]
I would trace the variable in question. It appears it may be aligned incorrectly. I am not sure why it would be peculiar to your environment though. The ABI should be identical.

**Comment by directhex** [comment]
```

**Comment by nealef** [comment]
Is this master? Line 8432 of method-to-ir.c in master looks like:

```
                                /*
                                 * We are compiling a call to a
                                 * generic method from shared code,
                                 * which means that we have to look up . <---- :8432
                                 * the method in the rgctx and do an
                                 * indirect call.
                                 */
```

**Comment by directhex** [comment]
No, it's 5.12, which I'm trying to do the integration work on Debian with.

8432 is https://salsa.debian.org/dotnet-team/mono/blob/master/mono/mini/method-to-ir.c#L8432

**Comment by nealef** [comment]
Where is the implementation of ```m_classgenericinst_get_generic_class```?

**Comment by directhex** [comment]
I think you might have mis-pasted the method name there, Neale?

**Comment by directhex** [comment]
BTW I did an -O0 build, in theory.

```
(sid_s390x-dchroot)directhex@zelenka:~/mono-5.12.0.309+dfsg/mcs/docs$ MONO_PATH="./../class/lib/net_4_x:$MONO_PATH" gdb /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono-sgen
GNU gdb (Debian 8.1-4) 8.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "s390x-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono-sgen...done.
(gdb) dir ../../mono/metadata
Source directories searched: /home/directhex/mono-5.12.0.309+dfsg/mcs/docs/../../mono/metadata:$cdir:$cwd
(gdb) start --debug ./../class/lib/net_4_x/mdoc.exe --debug assemble -o cs-errors -f error cs-errors.config
Temporary breakpoint 1 at 0x2d280: file main.c, line 394.
Starting program: /home/directhex/mono-5.12.0.309+dfsg/mono/mini/mono-sgen --debug ./../class/lib/net_4_x/mdoc.exe --debug assemble -o cs-errors -f error cs-errors.config
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/s390x-linux-gnu/libthread_db.so.1".

Temporary breakpoint 1, main (argc=10, argv=0x3fffffff4a8) at main.c:394
394     main.c: No such file or directory.
(gdb) c
Continuing.
[New Thread 0x3fffd3ff910 (LWP 20865)]
[New Thread 0x3fffb77b910 (LWP 20866)]

Thread 1 "mono-sgen" received signal SIGSEGV, Segmentation fault.
mono_metadata_get_inflated_signature (sig=sig@entry=0x1007bc870, context=context@entry=0x3f3851ec00000000) at metadata.c:3079
3079            helper.sig = sig;
(gdb) bt
#0  mono_metadata_get_inflated_signature (sig=sig@entry=0x1007bc870, context=context@entry=0x3f3851ec00000000) at metadata.c:3079
#1  0x00000001001ce34a in mono_method_get_signature_checked (method=method@entry=0x10080cf90, image=0x100537330, image@entry=0x2, token=token@entry=167772359, context=0x3f3851ec00000000,
    context@entry=<error reading variable: value has been optimized out>, error=error@entry=0x1006ec6f8) at loader.c:733
#2  0x00000001000811b0 in mono_method_to_ir (cfg=cfg@entry=0x1006ec270, method=method@entry=0x10080cf50, start_bblock=<optimized out>, start_bblock@entry=0x0, end_bblock=<optimized out>, end_bblock@entry=0x0, return_var=return_var@entry=0x0,
    inline_args=<optimized out>, inline_offset=<optimized out>, is_virtual_call=<optimized out>) at method-to-ir.c:8432
#3  0x000000010012d4d8 in mini_method_compile (method=method@entry=0x10080cf50, opts=opts@entry=370223487, domain=domain@entry=0x1004729f0, flags=flags@entry=JIT_FLAG_RUN_CCTORS, parts=parts@entry=0, aot_method_index=<optimized out>) at mini.c:3455
#4  0x000000010012f69e in mono_jit_compile_method_inner (method=method@entry=0x10080cf50, target_domain=target_domain@entry=0x1004729f0, opt=opt@entry=370223487, error=0x3ffffffd5a0, error@entry=<error reading variable: value has been optimized out>)
    at mini.c:4190
#5  0x0000000100032aa4 in mono_jit_compile_method_with_opt (method=method@entry=0x10080cf50, opt=<optimized out>, jit_only=jit_only@entry=0, error=<optimized out>, error@entry=0x3ffffffd5a0) at mini-runtime.c:2150
#6  0x0000000100033a36 in mono_jit_compile_method (method=method@entry=0x10080cf50, error=error@entry=0x3ffffffd5a0) at mini-runtime.c:2196
#7  0x00000001000cff3e in common_call_trampoline (regs=regs@entry=0x3ffffffd710, code=code@entry=0x3fffddc1598 "\022\"\300\204", m=m@entry=0x10080cf50, vt=vt@entry=0x0, vtable_slot=<optimized out>, vtable_slot@entry=0x0, error=0x3ffffffd5a0)
    at mini-trampolines.c:736
#8  0x00000001000d0844 in mono_magic_trampoline (regs=0x3ffffffd710, code=0x3fffddc1598 "\022\"\300\204", arg=0x10080cf50, tramp=<optimized out>) at mini-trampolines.c:867
#9  0x000003fffdfe60e2 in ?? ()
#10 0x000003fffddc1598 in ?? ()
#11 0x000003fffddc105c in ?? ()
#12 0x000003fffddc0e62 in ?? ()
#13 0x000003fffddc07f6 in ?? ()
#14 0x000003fffddc0682 in ?? ()
#15 0x000003fffddc03a2 in ?? ()
#16 0x000003fffddf1a12 in ?? ()
#17 0x000003fffddf116a in ?? ()
#18 0x000003fffddf0dfa in ?? ()
#19 0x000003fffddf0c22 in ?? ()
#20 0x000003fffddf0b6a in ?? ()
#21 0x000003fffddf09b0 in ?? ()
#22 0x000003fffddf045a in ?? ()
#23 0x000003fffdedffb6 in ?? ()
#24 0x000003fffdecaf14 in ?? ()
#25 0x000003fffdfd104e in ?? ()
#26 0x000003fffdfcf70a in ?? ()
PC not saved
```

**Comment by nealef** [comment]
> I think you might have mis-pasted the method name there, Neale?

No, if you follow the chain of calls you end up in class-accessors.c and that call appears:

```
/*
* mono_class_try_get_generic_class:
*
*   Return the MonoGenericClass if @klass is a ginst, NULL otherwise
*/
MonoGenericClass*
mono_class_try_get_generic_class (MonoClass *klass)
{
        if (mono_class_is_ginst (klass))
                return m_classgenericinst_get_generic_class ((MonoClassGenericInst*)klass);
        return NULL;
}
```

**Comment by nealef** [comment]
It's defined as part of this macro:

```
/* If MonoClass definition is hidden, just declare the getters.
 * Otherwise, define them as static inline functions.
 */
#ifdef MONO_CLASS_DEF_PRIVATE
#define MONO_CLASS_GETTER(funcname, rettype, optref, argtype, fieldname) rettype funcname (argtype *klass);
#else
#define MONO_CLASS_GETTER(funcname, rettype, optref, argtype, fieldname) static inline rettype funcname (argtype *klass) { return optref klass-> fieldname ; }
#endif
```

**Comment by nealef** [comment]
So I'd check if the value of klass i think it's has the 0x3f... value that I don't think is valid. Next see where that value is set.

**Comment by directhex** [comment]
I worked with @nealef on this. My specific issue with MDOC only happens when Mono is built with GCC 6 or above, so needs GCC bisecting to figure out the specific change (whether a bug in GCC, or a bug in Mono which is exposed by a change in GCC)

**Comment by nealef** [comment]
It works fine with CentOS 7.5.1804 clone using the devtoolset-7 gcc (7.1.3).

**Comment by directhex** [comment]
Some combination of GCC flags, libc, etc.

The one thing we can be sure of is it's not the kernel, because you saw it in a chroot on a SUSE kernel

**Comment by nealef** [comment]
Fixes required to get building under Ubuntu 18.04

* PIE issue: Add ```-no-pie``` to CFLAGS/CXXFLAGS when running autogen.sh. This is a circumvention that prevents a problem of a null variable being passed from ```method-to-ir``` to ```mono_method_get_signature_checked```

* MonoContext: Conditional compilation based on GLIBC level as struct ucontext has changed to ucontext_t
```
--- a/mono/utils/mono-context.h
+++ b/mono/utils/mono-context.h
@@ -865,7 +865,13 @@ typedef struct {
 
 #define MONO_ARCH_HAS_MONO_CONTEXT 1
 
+#include <sys/ucontext.h>
+
+#if __GLIBC_PREREQQ(2, 27)
+typedef ucontext_t MonoContext;
+#else
 typedef struct ucontext MonoContext;
+#endif
 
 #define MONO_CONTEXT_SET_IP(ctx,ip)                                    \
        do {    
```

* Successful running of tests:
1. Backport ```cd70e9cf2adc03681dbb919d97063d8ad84b0a4c``` for tailcall tests
2. Fix mono/tests/Makefile.am to use langversion of 7.2 as rosyln doesn't work on big endian systems.
```
--- a/mono/tests/Makefile.am
+++ b/mono/tests/Makefile.am
@@ -207,7 +207,7 @@ MCS_NO_UNSAFE = $(TOOLS_RUNTIME) $(CSC) -debug:portable \
        -noconfig -nologo \
        -nowarn:0162 -nowarn:0168 -nowarn:0219 -nowarn:0414 -nowarn:0618 \
        -nowarn:0169 -nowarn:1690 -nowarn:0649 -nowarn:0612 -nowarn:3021 \
-       -nowarn:0197 -langversion:7.3 $(PROFILE_MCS_FLAGS)
+       -nowarn:0197 -langversion:7.2 $(PROFILE_MCS_FLAGS)
 MCS_NO_LIB = $(MCS_NO_UNSAFE) -unsafe
 
 MCS = $(MCS_NO_LIB)
```

* Backport ```6d07dd66f1412b8f7eaadbcd9d660859d4227954``` otherwise build of mcs won't even begin

**Comment by nealef** [comment]
For the 7.3->7.2 issue what is the best approach?
* ```if S390X``` (and AIX/PPC64) within the ```mono/tests/Makefile.am``` 
* In ```configure.ac``` add a ```BIG_ENDIAN``` variable (that could also be used for defaulting to ```mcs``` instead of ```roslyn```
* Something else?
Then using whatever approach set ```LANGVERSION``` to 7.3 or 7.2 then use that variable:
```
-nowarn:0197 -langversion:$(LANGVERSION) $(PROFILE_MCS_FLAGS)
```

**Comment by nealef** [comment]
PR #12272 mono-context.h patch. If this acceptable it should be backported to 2018-08.

**Comment by akoeplinger** [comment]
Backported #12272 to 2018-10 which will be our next release, 2018-08 is already closed.

**Comment by akoeplinger** [comment]
@nealef I think only the PIE issue is left, should we extract that one into a new issue?

**Comment by nealef** [comment]
Didn't that get fixed? Or are you looking at backporting it?

**Comment by akoeplinger** [comment]
I was just going through some open issue, do you happen to know the PR which fixed it?

**Comment by nealef** [comment]
37d6cde0ee0ea6b788cd367040609bf741aa3269 ?

**Comment by akoeplinger** [comment]
Hmm, that commit is from December 2016 so should long be in master.  Just to make sure we're on the same page, I'm talking about what you mentioned in https://github.com/mono/mono/issues/9009#issuecomment-451269331

> * PIE issue: Add `-no-pie` to CFLAGS/CXXFLAGS when running autogen.sh. This is a circumvention that prevents a problem of a null variable being passed from `method-to-ir` to `mono_method_get_signature_checked`

**Comment by nealef** [comment]
But last time you said you got a clean build with Ubuntu I don't think I had made any other changes. Is it failing now?

**Comment by akoeplinger** [comment]
It builds on [Jenkins](https://jenkins.mono-project.com/job/test-mono-mainline-community/label=centos-s390x/) and since you don't need any additional changes I'll just assume @directhex will upstream anything he needs separately :)

**Comment by sharkcz** [comment]
> Fixes required to get building under Ubuntu 18.04
> 
>     * PIE issue: Add `-no-pie` to CFLAGS/CXXFLAGS when running autogen.sh. This is a circumvention that prevents a problem of a null variable being passed from `method-to-ir` to `mono_method_get_signature_checked`

this could be the same issue we are tracking in https://bugzilla.redhat.com/show_bug.cgi?id=1397948

**Comment by jaykrell** [comment]
Maybe remove the S390X fast TLS support??

**Comment by sharkcz** [comment]
For the record - starting with https://src.fedoraproject.org/rpms/mono/c/cb2069e8baa6814a1094b4ac700047ee42139f16?branch=master we have a build of 5.18.1 in Fedora for all our arches (https://koji.fedoraproject.org/koji/buildinfo?buildID=1239640)

**Comment by nealef** [comment]
I think the -fPIE/-fpie is a problem. I think it revealed issues with the TLS fastpath that were addressed by its simplification. Coupled with the conditional logic when created the context related structure I think these two are all that was required.

**Comment by mgorse** [comment]
>     * MonoContext: Conditional compilation based on GLIBC level as struct ucontext has changed to ucontext_t
> 
> 
> ```
> --- a/mono/utils/mono-context.h
> +++ b/mono/utils/mono-context.h
> @@ -865,7 +865,13 @@ typedef struct {
>  
>  #define MONO_ARCH_HAS_MONO_CONTEXT 1
>  
> +#include <sys/ucontext.h>
> +
> +#if __GLIBC_PREREQ(2, 27)
> +typedef ucontext_t MonoContext;
> +#else
>  typedef struct ucontext MonoContext;
> +#endif
>  
>  #define MONO_CONTEXT_SET_IP(ctx,ip)                                    \
>         do {    

I think that we need this for glibc 2.26 as well. I'm getting this build failure on SLE 15, which has 2.26.
